src/common/cvt.cpp

```diff
@@ -1033,6 +1047,928 @@ void CVT_string_to_datetime(const dsc* desc,
+				if (hours > 23)
+				{
+					cb->err(Arg::Gds(isc_value_for_pattern_is_out_of_range) <<
+						string(pattern.data(), pattern.length()) << Arg::Num(0) << Arg::Num(23));
+				}
+
+				outTimes.tm_hour = hours;
+				return;
+			}
+			break;
+
+		case 'S':
+			if (pattern == "SS")
+			{
+				int seconds = parse_string_to_get_int(str, strLength, strOffset, 2);
+				if (seconds > 59)
+				{
+					cb->err(Arg::Gds(isc_value_for_pattern_is_out_of_range) <<
+						string(pattern.data(), pattern.length()) << Arg::Num(0) << Arg::Num(59));
+				}
+
+				outTimes.tm_sec = seconds;
+				return;
+			}
+			else if (pattern == "SSSSS")
+			{
+				constexpr int maximumSecondsInDay = 23 * 60 * 60 + 59 * 60 + 59;
+
+				int secondsInDay = parse_string_to_get_int(str, strLength, strOffset, 5);
+				if (secondsInDay > maximumSecondsInDay)
+				{
+					cb->err(Arg::Gds(isc_value_for_pattern_is_out_of_range) <<
+						string(pattern.data(), pattern.length()) << Arg::Num(0) << Arg::Num(maximumSecondsInDay));
+				}
+
+				int hours = secondsInDay / 24;
+				int minutes = secondsInDay / 60 - hours * 60;
+				int seconds = secondsInDay - minutes * 60 - hours * 60 * 60;
+
+				outTimes.tm_hour = hours;
+				outTimes.tm_min = minutes;
+				outTimes.tm_sec = seconds;
+				return;
+			}
+			break;
+
+		case 'F':
+			if (!strncmp(pattern.data(), "FF", pattern.length() - 1))
+			{
+				int number = pattern.back() - '0';
+				if (number < 1 || number > -ISC_TIME_SECONDS_PRECISION_SCALE)
+				{
+					invalidPatternException(pattern, cb);
+				}
+
+				const int fractions = parse_string_to_get_int(str, strLength, strOffset, number);
+				outFractions = fractions * pow(10, -ISC_TIME_SECONDS_PRECISION_SCALE - number);
+				return;
+			}
+			break;
+
+		case 'T':
+			if (pattern == "TZH")
+			{
+				if (previousPattern == "TZM")
+				{
+					outTimezoneInMinutes += sign(outTimezoneInMinutes) *
+						parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset) * 60;
+				}
+				else
+					outTimezoneInMinutes = parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset, true) * 60;
+				return;
+			}
+			else if (pattern == "TZM")
+			{
+				if (previousPattern == "TZH")
+				{
+					outTimezoneInMinutes += sign(outTimezoneInMinutes) *
+						parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset);
+				}
+				else
+					outTimezoneInMinutes = parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset, true);
+				return;
+			}
+			break;
+	}
+
+	invalidPatternException(pattern, cb);
+}
+
+
+ISC_TIMESTAMP_TZ CVT_string_to_format_datetime(const dsc* desc, const Firebird::string& format, Firebird::Callbacks* cb)
+{
+	if (!DTYPE_IS_TEXT(desc->dsc_dtype))
+		cb->err(Arg::Gds(isc_invalid_data_type_for_date_format));
+
+	if (format.isEmpty())
+		cb->err(Arg::Gds(isc_sysf_invalid_null_empty) << Arg::Str(STRINGIZE(format)));
+
+	auto invalidPatternException = [](std::string_view pattern, Callbacks* cb)
```

Comment:
`std::string_view` is C++17 feature, not allowed to use in current Firebird code.
But since this should not go to Firebird 5, I will propose adoption of C++17 for Firebird 6 code.

Reply:
I see other people agree to move to C++ 17, so I switch back static to inline variables.